transaction/std/src/memo_builder/gift_code_funding_memo_builder.rs

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Defines the Memo Builder for the gift code funding memo (0x0201)
+//! specified in MCIP #32
+
+use crate::{
+    memo::{GiftCodeFundingMemo, UnusedMemo},
+    MemoBuilder, ReservedDestination,
+};
+use mc_account_keys::PublicAddress;
+use mc_crypto_keys::RistrettoPublic;
+use mc_transaction_core::{Amount, MemoContext, MemoPayload, NewMemoError};
+
+/// There are three possible gift code memo types specified in MCIP #32
+/// | Memo type bytes | Name                        |
+/// | -----------     | -----------                 |
+/// |    0x0002       | Gift Code Sender Memo       |
+/// | -->0x0201<--    | Gift Code Funding Memo      |
+/// |    0x0202       | Gift Code Cancellation Memo |
+/// This memo builder builds a gift code funding memo (0x0201). When a gift
+/// code is funded, the amount of the gift code is sent to a TxOut at the
+/// Sender's reserved gift code subaddress and a second (potentially zero
+/// valued) change TxOut is sent to the sender's reserved change subaddress
+/// with the gift code funding memo attached. The funding memo will include
+/// the first 4 bytes of the hash of the gift code TxOut sent to the
+/// sender's reserved gift code subaddress and 60 bytes for an optional
+/// utf-8 memo.
+#[derive(Clone, Debug)]
+pub struct GiftCodeFundingMemoBuilder {
+    // Utf-8 note within the memo that can be up to 60 bytes long
+    note: String,
+    // TxOut Public Key of the gift code TxOut sent to the gift code subaddress
+    gift_code_tx_out_public_key: Option<RistrettoPublic>,
+    // Whether we've already written the change memo
+    wrote_change_memo: bool,
+    // Whether we've already written the gift code TxOut
+    wrote_destination_memo: bool,
```

Comment:
Yeah I had thought about this. While this is true it's implied by being none, if someone needs to maintain this who isn't an original author of it comes along, this makes it more explicit/readable. For the maintainability aspect, I'd vote for retaining it?

Reply:
I don't think I agree with that, but don't feel strongly. I think having less `if` statements (one of which will never actually be true due to this redundancy) is more maintainable.

Reply:
@iamalwaysuncomfortable generally I would argue that less mutable state is more maintainable

if you want to make it more explicit / readable, because you think, `self.wrote_destination_memo` is more readable than `self.tx_out_public_key.is_some()`, another thing you could do is make this a private function instead, so its `self.wrote_destination_memo()` ?

then you don't have an extra variable that someone could forget to update while refactoring this.